File: python-selenium-pytest-bdd/src/saucedemo_uitests/support/polling.py
"""Polls an arbitrary zero-arg callable until it returns the expected value or a timeout elapses - the
Python/Selenium equivalent of the C# sibling's NUnit `Assert.That(() => ..., Is.EqualTo(x).After(ms, ms))`
polling assertions used for state that updates asynchronously (e.g. a cart badge re-rendering after a
click).

Deliberately a plain time.monotonic() loop with a direct `==` check rather than built on
`WebDriverWait.until` (as `pages/base_page.py` is): `until` stops as soon as its callback returns
anything falsy-looking as "not ready" - which would misfire for a legitimately expected falsy value (e.g.
asserting a count of exactly 0), so a direct equality check is the correct general-purpose contract here,
not just a stylistic difference.
"""

import time
from collections.abc import Callable
from typing import TypeVar

_T = TypeVar("_T")

_DEFAULT_TIMEOUT_SECONDS = 5.0
_DEFAULT_POLL_SECONDS = 0.25


def assert_eventually(
    read: Callable[[], _T],
    expected: _T,
    description: str,
    *,
    timeout_seconds: float = _DEFAULT_TIMEOUT_SECONDS,
    poll_seconds: float = _DEFAULT_POLL_SECONDS,
) -> None:
    deadline = time.monotonic() + timeout_seconds
    actual = read()

    while actual != expected and time.monotonic() < deadline:
        time.sleep(poll_seconds)
        actual = read()

    assert actual == expected, f"{description}: expected {expected!r}, got {actual!r} after polling."

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/support/Polling.java
package com.saucedemo.uitests.support;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.Duration;
import java.util.function.Supplier;

/**
 * Polls a supplier until it returns the expected value or a timeout elapses - for state that updates
 * asynchronously (e.g. a cart badge re-rendering after a click).
 *
 * <p>Deliberately a plain loop with a direct equality check rather than something built on
 * {@code WebDriverWait.until}: {@code until} stops as soon as its callback returns anything falsy-looking
 * and treats that as "not ready", which would misfire for a legitimately expected falsy value (asserting
 * a cart count of exactly 0, say). A direct equality check is the correct general-purpose contract here,
 * not just a stylistic difference.
 */
public final class Polling {

    private static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(5);
    private static final Duration DEFAULT_POLL_INTERVAL = Duration.ofMillis(250);

    private Polling() {
    }

    public static <T> void assertEventually(Supplier<T> read, T expected, String description) {
        assertEventually(read, expected, description, DEFAULT_TIMEOUT);
    }

    public static <T> void assertEventually(Supplier<T> read, T expected, String description, Duration timeout) {
        long deadline = System.nanoTime() + timeout.toNanos();
        T actual = read.get();

        while (!expected.equals(actual) && System.nanoTime() < deadline) {
            sleep(DEFAULT_POLL_INTERVAL);
            actual = read.get();
        }

        assertThat(actual).as("%s (after polling for %s)", description, timeout).isEqualTo(expected);
    }

    private static void sleep(Duration duration) {
        try {
            Thread.sleep(duration.toMillis());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Interrupted while polling.", e);
        }
    }
}
